url2a908001cd_malictus_klang_tgz-pJ8-malictus_klang_test_PrimitiveTestJ8/src/malictus/klang/chunk/PNGFileChunk.java

```diff
@@ -0,0 +1,192 @@
+			//insert name + size
+			KlangUtil.insertIntoFile(new byte[8], this.getParentFile().getFile(), startingplace);
+			raff.seek(startingplace);
+			//write size
+			Int4ByteUnsignedBE size = new Int4ByteUnsignedBE();
+			size.setValue(theFile.length());
+			size.writeValueToFile(raff);	
+			//write name
+			FourCC fcc = new FourCC(name);
+			fcc.writeValueToFile(raff);
+			//write CRC
+			raff.seek(raff.getFilePointer() + theFile.length());
+			long val = KlangUtil.getCRCFor(this.getParentFile().getFile(), startingplace + 4, raff.getFilePointer());
+			Int4ByteUnsignedBE cr = new Int4ByteUnsignedBE();
+			cr.setValue(val);
+			cr.writeValueToFile(raff);
+			//now update and reparse
+			((EditableFileBase)this.getParentFile()).reparseFile(raff);
+			this.chunkJustChanged(raff);
+		} catch (BadValueException err) {
+			throw new IOException(KlangConstants.ERROR_GENERAL_FILE);
+		}
+	}
+	
+	public void addChunkFromArray(byte[] bytes, String name, int position, RandomAccessFile raff) throws IOException {
+		if (!(this.getParentFile() instanceof EditableFileBase)) {
+			throw new IOException(KlangConstants.ERROR_GENERAL_FILE);
+		}
+		long newsize = (this.endPosition - this.startPosition) + bytes.length;
+		if (newsize < 0) {
+			throw new IOException(KlangConstants.ERROR_GENERAL_FILE);
+		}
+		try {
+			//warn parent chunks
+			this.chunkIsAboutToChangeSize(newsize, raff);
+			//now make the addition
+			long startingplace = 0;
+			if (position > this.getSubChunks().size() + 2) {
+				position = this.getSubChunks().size() + 1;
+			}
+			if (position == this.getSubChunks().size() + 1) {
+				if (this.getSubChunks().size() == 0) {
+					startingplace = this.endPosition;
+				} else {
+					startingplace = this.getSubChunks().get(position - 2).getEndPosition();
+				}
+			} else {
+				startingplace = this.getSubChunks().get(position - 1).getStartPosition();
+			}
+			//insert CRC
+			KlangUtil.insertIntoFile(new byte[4], this.getParentFile().getFile(), startingplace);
+			//insert file
+			KlangUtil.insertIntoFile(bytes, this.getParentFile().getFile(), startingplace);
+			//insert name + size
+			KlangUtil.insertIntoFile(new byte[8], this.getParentFile().getFile(), startingplace);
+			raff.seek(startingplace);
+			//write size
+			Int4ByteUnsignedBE size = new Int4ByteUnsignedBE();
+			size.setValue(bytes.length);
+			size.writeValueToFile(raff);	
+			//write name
+			FourCC fcc = new FourCC(name);
+			fcc.writeValueToFile(raff);
+			//write CRC
+			raff.seek(raff.getFilePointer() + bytes.length);
+			long val = KlangUtil.getCRCFor(this.getParentFile().getFile(), startingplace + 4, raff.getFilePointer());
+			Int4ByteUnsignedBE cr = new Int4ByteUnsignedBE();
+			cr.setValue(val);
+			cr.writeValueToFile(raff);
+			//now update and reparse
+			((EditableFileBase)this.getParentFile()).reparseFile(raff);
+			this.chunkJustChanged(raff);
+		} catch (BadValueException err) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)